network/ClientNetworking.cpp

```diff
@@ -127,13 +128,23 @@ ClientNetworking::ClientNetworking() :
     m_io_service(),
     m_socket(m_io_service),
     m_incoming_messages(m_mutex),
-    m_connected(false),
-    m_cancel_retries(false)
+    m_rx_connected(false),
+    m_tx_connected(false)
 {}
 
-bool ClientNetworking::Connected() const {
+bool ClientNetworking::IsConnected() const {
     boost::mutex::scoped_lock lock(m_mutex);
-    return m_connected;
+    return m_rx_connected && m_tx_connected;
+}
+
+bool ClientNetworking::IsRxConnected() const {
+    boost::mutex::scoped_lock lock(m_mutex);
```

Comment:
Yes synchronisation issues occur from very low level effects right through to higher levels.

I chose the example to pre-emptively point out that the network thread only writing and the main thread only reading is not safe.  I picked the lowest level that occurred to me.

At a higher level, outside of the mutex the compiler/OS and/or the CPU are allowed to perform instruction re-ordering and assume that there is a single thread of execution.  They would be allowed to check the `m_rx_connected` and `m_txt_connected` in either order, which would be not good if the writing thread updated them in a different order from the reading thread. 

If you wish to further titillate yourself, look up synchronization primitives for the processor of your choice.

Can either of you suggest a better comment for `m_mutex` so that it is clear to more people, why the mutex is being used?

The current comment is,
````
    // m_mutex guards m_incoming_message, m_rx_connected and m_tx_connected which are written by
    // the networking thread and read by the main thread to check incoming messages and connection
    // status.
````